bar_app/serializers.py

```diff
@@ -0,0 +1,169 @@
+from django.contrib.auth.models import User
+from .models import *
+
+class UserSerializer(serializers.ModelSerializer):
+    """
+    Serializer for User model.
+    """
+    class Meta:
+        model = User
+        fields = ["username", "password"]
+        extra_kwargs = {"password": {"write_only": True}}
+
+class ReferenceSerializer(serializers.ModelSerializer):
+    """
+    Serializer for Reference model.
+    """
+    availability = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Reference
+        fields = "__all__"
+        read_only_fields = ["id"]
+
+    def validate_ref(self, value):
+        """
+        Check if the ref exists in the database.
+        """
+        if Reference.objects.filter(ref=value).exists():
+            raise serializers.ValidationError("This reference already exists.")
+        return value
+
+    def validate_name(self, value):
+        """
+        Validate the data for creating a new reference.
+        """
+        if not value.strip():
+            raise serializers.ValidationError("Name field cannot be empty.")
+
+        return value
+
+    def get_availability(self, instance):
+        stocks = Stock.objects.filter(reference=instance)
+
+        if not stocks.exists():
+            return "outofstock"
+
+        for stock in stocks:
+            if stock.stock > 0:
+                return "available"
+        return "outofstock"
+
+class BarSerializer(serializers.ModelSerializer):
+    """
+    Serializer for Bar model.
+    """
+    class Meta:
+        model = Bar
+        fields = "__all__"
+        read_only_fields = ["id"]
+
+    def validate_name(self, value):
+        """
+        Check if the name field is not empty.
+        """
+        if not value.strip():
+            raise serializers.ValidationError("Name field cannot be empty.")
+        return value
+
+    def validate_id(self, value):
+        """
+        Check if the provided stock exists in the database.
+        """
+        if self.instance:
+            if not Bar.objects.filter(id=value).exists():
+                raise serializers.ValidationError("Invalid bar id.")
+        return value
+
+class StockSerializer(serializers.ModelSerializer):
+    """
+    Serializer for Stock model.
+    """
+    class Meta:
+        model = Stock
+        fields = "__all__"
+        read_only_fields = ["id"]
+
+class StatisticsSerializer(serializers.Serializer):
+    """
+    Serializer for statistics data.
+    """
+    all_stocks = serializers.SerializerMethodField()
+    miss_at_least_one = serializers.SerializerMethodField()
+
+    def get_all_stocks(self):
+        """
+        Retrieve a list of bars that have all references in stock.
+
+        Returns:
+            dict: A dictionary containing the description and list of bars with all stocks.
+        """
+        bars_with_all_stocks = []
+        bars = Bar.objects.all()
+
+        for bar in bars:
+            stocks = Stock.objects.filter(bar=bar).all()
```

Comment:
you should (virtually) never do a database hit inside a loop. use only django ORM